crates/runtime/src/instance/allocator.rs

```diff
@@ -104,104 +282,160 @@ pub unsafe trait InstanceAllocator {
     ///
     /// Note that the returned instance must still have `.initialize(..)` called
     /// on it to complete the instantiation process.
-    fn allocate(&self, mut req: InstanceAllocationRequest) -> Result<InstanceHandle> {
-        let index = self.allocate_index(&req)?;
-        let module = req.runtime_info.module();
-        let mut memories =
-            PrimaryMap::with_capacity(module.memory_plans.len() - module.num_imported_memories);
-        let mut tables =
-            PrimaryMap::with_capacity(module.table_plans.len() - module.num_imported_tables);
+    ///
+    /// # Unsafety
+    ///
+    /// The request's associated module, memories, tables, and vmctx must have
+    /// already have been validated by `Self::validate_module`.
+    unsafe fn allocate_module(
+        &self,
+        mut request: InstanceAllocationRequest,
+    ) -> Result<InstanceHandle> {
+        let module = request.runtime_info.module();
+
+        #[cfg(debug_assertions)]
+        InstanceAllocatorImpl::validate_module_impl(self, module, request.runtime_info.offsets())
+            .expect("module should have already been validated before allocation");
+
+        self.increment_core_instance_count()?;
+
+        let num_defined_memories = module.memory_plans.len() - module.num_imported_memories;
+        let mut memories = PrimaryMap::with_capacity(num_defined_memories);
+
+        let num_defined_tables = module.table_plans.len() - module.num_imported_tables;
+        let mut tables = PrimaryMap::with_capacity(num_defined_tables);
 
         let result = self
-            .allocate_memories(index, &mut req, &mut memories)
-            .and_then(|()| self.allocate_tables(index, &mut req, &mut tables));
+            .allocate_memories(&mut request, &mut memories)
+            .and_then(|()| self.allocate_tables(&mut request, &mut tables));
         if let Err(e) = result {
-            self.deallocate_memories(index, &mut memories);
-            self.deallocate_tables(index, &mut tables);
-            self.deallocate_index(index);
+            self.deallocate_memories(&mut memories);
+            self.deallocate_tables(&mut tables);
+            self.decrement_core_instance_count();
```

Comment:
Note from Chris on zoom: can we RAII this stuff?

Reply:
or we can use an immediately-invoked closure so that we can `?` inside it but still have only a single place for clean up on error

Reply:
Note: I am going to avoid an RAII type because

1. we would need two types, one for core module instantiation and one for component instantiation, and
2. only the error paths would be able to use these, not the success paths, because the RAII type would need to wrap `self` here, which isn't feasible once the instance is returned.

Therefore, I'll do the immediately-invoked closure thing instead.